game-core/src/main/java/games/strategy/triplea/attachments/TerritoryAttachment.java

```diff
@@ -614,11 +614,11 @@ private void resetBlockadeZone() {
    * Returns the collection of territories that make up the convoy route containing the specified
    * territory or {@code null} if the specified territory is not part of a convoy route.
    */
-  public static Set<Territory> getWhatTerritoriesThisIsUsedInConvoysFor(
+  public static Collection<Territory> getWhatTerritoriesThisIsUsedInConvoysFor(
```

Comment:
Hmm, defining Set as return type usually implies that there are no duplicates, so I'd actually like to keep the interface defined as Set

Reply:
The javadoc states collection:
> Returns the collection of territories that make up the convoy route containing the specified
   * territory 

We probably should update L623 as well:
```
 final Set<Territory> territories = new HashSet<>();
```
The fact no duplicates are returned is an attribute of the implementation. We've a problem where we over-specify 'Set' or 'List' in many places where we rely on nor need either API. Looking back at BattleTracker, it looks like a good bit of this over-specification came from the original 2003 code, are we continuing bad habits?

You can also have a List containing all unique elements as well, the fact that you get set does not mean that no other collection type will not also include duplicates. IE: `Set` type is not the only collection type or situation where you may see all unique values.

Reply:
I updated the javadoc as it also stated that the return value was nullable: 5831782402ad94ffe26756f66d4a88e9d8520605

Reply:
Checking for any 3rd party references on best practices here, there's not too much. The examples I have found do encourage returning `Collection` type:


- https://www.codejava.net/java-core/collections/18-java-collections-and-generics-best-practices#PreferInterface
- https://www.javaguides.net/2018/06/java-collection-framework-best-practices.html

Examples of in-line code assign to `List`, but those examples are for local scoped variables. Given we are setting a contract, the more generic of interface type seems to be the rule of thumb.

Reply:
I personally prefer to return Set since the caller then knows you have unique elements.

Reply:
IDK, arguably if a caller can generalize to a `Collection`, then they should. IE: don't depend on a `Set` return type if you all you need is an iterator. In that case if the callee changes, the caller can be unchanged. 

I can see some merits to both sides. My concerns for this codebase:
- continues a pattern of using concrete types and/or Set/List when all we need is Collection. There is a mix, most of the v1 code uses the older types, looks like when Chris got involved much was made to depend on Collection
- It's not often that a caller actually cares about uniqueness, it pushes details that we maybe should not be concerned about. #bikeshedding  It's very similar I think to the underscore naming pattern, it pushes a lot of unnecessary details higher in importance than is actually necessary, by the time you're done with the ceremonies you can then worry about the actual business code
- We probably should be using `Collection` as internal type. It's the implementation that defines the uniqueness and ordering behavior, not the type. IE:
```
Collection<String> strings = new HashSet<>();
```
If we then have 'getters', it's going to be odd that some methods return `Collection` and others return `Set`.